src/ua_types.c

```diff
@@ -1010,13 +1011,37 @@ copyStructure(const void *src, void *dst, const UA_DataType *type) {
     for(size_t i = 0; i < type->membersSize; ++i) {
         const UA_DataTypeMember *m= &type->members[i];
         const UA_DataType *mt = &typelists[!m->namespaceZero][m->memberTypeIndex];
-        if(!m->isArray) {
+        if (m->isOptional){
             ptrs += m->padding;
             ptrd += m->padding;
-            retval |= copyJumpTable[mt->typeKind]((const void*)ptrs, (void*)ptrd, mt);
+            if((!m->isArray && (*(void* const*)ptrs != NULL)) ||
```

Comment:
We have optional and array and combinations thereof.
So four different cases.

You check for m->isArray in 5 different places.
That makes the code hard to read.
It should be enough to test this just once.